src/deploy/extensions/secrets.ts

```diff
@@ -0,0 +1,232 @@
+    await handleSecretParamForCreate(s, i, nonInteractive);
+  }
+}
+
+async function handleSecretsUpdateInstance(
+  i: InstanceSpec,
+  prevSpec: InstanceSpec,
+  nonInteractive: boolean
+) {
+  const extensionVersion = await getExtensionVersion(i);
+  const prevExtensionVersion = await getExtensionVersion(prevSpec);
+  const secretParams = secretsInSpec(extensionVersion.spec);
+  for (const s of secretParams) {
+    // If this was previously a secret param & was set, treat this as an update
+    const prevParam = prevExtensionVersion.spec.params.find((p) => p.param === s.param);
+    if (prevParam?.type === ParamType.SECRET && prevSpec.params[prevParam?.param]) {
+      await handleSecretParamForUpdate(s, i, prevSpec.params[prevParam?.param], nonInteractive);
+    } else {
+      // Otherwise, this is a new secret param
+      await handleSecretParamForCreate(s, i, nonInteractive);
+    }
+  }
+}
+
+async function handleSecretParamForCreate(
+  secretParam: Param,
+  i: InstanceSpec,
+  nonInteractive: boolean
+): Promise<void> {
+  const providedValue = i.params[secretParam.param];
+  if (!providedValue) {
+    return;
+  }
+  // First, check that param is well formed.
+  const [, projectId, , secretName, , version] = providedValue.split("/");
+  if (!projectId || !secretName || !version) {
+    throw new FirebaseError(
+      `${clc.bold(i.instanceId)}: Found '${providedValue}' for secret param ${
+        secretParam.param
+      }, but expected a secret version.`
+    );
+  }
+  // Then, go get all the info about the current state of the secret.
+  const secretInfo = await getSecretInfo(projectId, secretName, version);
+  // If the secret doesn't exist, prompt the user for a value, create it, and label it.
+  if (!secretInfo.secret) {
+    await promptForCreateSecret({
+      projectId,
+      secretName,
+      instanceId: i.instanceId,
+      secretParam,
+      nonInteractive,
+    });
+  } else if (!secretInfo.secretVersion) {
+    throw new FirebaseError(
+      `${clc.bold(i.instanceId)}: Found '${providedValue}' for secret param ${
+        secretParam.param
+      }. ` +
+        `projects/${projectId}/secrets/${secretName} exists, but version ${version} does not. ` +
+        `See more information about this secret at ${secretManager.secretManagerConsoleUri(
+          projectId
+        )}`
+    );
+  }
+  if (
+    !!secretInfo?.labels[secretUtils.SECRET_LABEL] &&
+    secretInfo.labels[secretUtils.SECRET_LABEL] !== i.instanceId
+  ) {
+    throw new FirebaseError(
+      `${clc.bold(i.instanceId)}: Found '${providedValue}' for secret param ${
+        secretParam.param
+      }. ` +
+        `projects/${projectId}/secrets/${secretName} is managed by a different extension instance (${
+          secretInfo.labels[secretUtils.SECRET_LABEL]
+        }), so reusing it here can lead to unexpected behavior. ` +
+        "Please choose a different name for this secret, and rerun this command."
+    );
+  }
```

Comment:
If secret/version exist and don't have the label we just use it, right? I think you still need to call `secretsUtils.grantFirexServiceAgentSecretAdminRole(secretInfo.secret)` on it so that we get all the necessary P4SA permissions. Similar in the update case.